nvbench/detail/entropy_criterion.cxx

```diff
@@ -0,0 +1,156 @@
+/*
+ *  Copyright 2023 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <nvbench/detail/entropy_criterion.cuh>
+#include <nvbench/types.cuh>
+
+#include <cmath>
+
+
+namespace nvbench::detail
+{
+
+entropy_criterion::entropy_criterion()
+  : stopping_criterion{"entropy"}
+{
+  m_freq_tracker.reserve(m_entropy_tracker.capacity() * 2);
+  m_ps.reserve(m_entropy_tracker.capacity() * 2);
+}
+
+void entropy_criterion::initialize(const criterion_params &params)
+{
+  m_params = params;
+  m_total_samples = 0;
+  m_total_cuda_time = 0.0;
+  m_entropy_tracker.clear();
+  m_freq_tracker.clear();
+
+  if (m_params.has_value("max-angle"))
+  {
+    m_max_angle = m_params.get_float64("max-angle");
+  }
+
+  if (m_params.has_value("min-r2"))
+  {
+    m_min_r2 = m_params.get_float64("min-r2");
+  }
+}
+
+nvbench::float64_t entropy_criterion::compute_entropy() 
+{
+  const std::size_t n = m_freq_tracker.size();
+  if (n == 0)
+  {
+    return 0.0;
+  }
+
+  m_ps.resize(n);
+  for (std::size_t i = 0; i < n; i++)
+  {
+    m_ps[i] = static_cast<nvbench::float64_t>(m_freq_tracker[i].second) /
+              static_cast<nvbench::float64_t>(m_total_samples);
+  }
+
+  nvbench::float64_t entropy{};
+  for (nvbench::float64_t p : m_ps)
+  {
+    entropy -= p * std::log2(p);
+  }
+
+  return entropy;
+}
+
+void entropy_criterion::add_measurement(nvbench::float64_t measurement)
+{
+  m_total_samples++;
+  m_total_cuda_time += measurement;
+
+  {
+    auto key = measurement;
+    constexpr bool bin_keys = false;
+
+    if (bin_keys) 
+    {
+      const auto resolution_us = 0.5;
+      const auto resulution_s = resolution_us / 1'000'000;
+      const auto epsilon = resulution_s * 2;
+      key = std::round(key / epsilon) * epsilon;
+    }
```

Comment:
Definitely good to have a binning option available to investigate. 
 
I think that the image example from the MR description for `cub.bench.scan.exclusive.sum::base` is a good example of something that I think might benefit from it. I'm not the most familiar with the specifics of cuda timing interface, however having worked with data from the perfmon hardware I am assuming it is based on, I can say that there is definitely a lower limit to the data granularity simply due to the polling rate on the hardware. But beyond even the practicality of measuring things this precise, I think it's worth asking if that data is useful to the user? Just something to think about.